### PR TITLE
Remove the injector

### PR DESCRIPTION
**Pull request by ty-everett**
This is a work in progress. I have removed the injector and all code references to it. However, certain classes were previously exported in multiple variants by using the injector to pass in—among a few other things—either the `Mainnet` or `Testnet` configurations from the `constants` object.

As of now, I have refrained from exporting those extra variants which has resulted in tests that rely on such things as `bip32.Testnet` to break. Without the use of the injector, from what I can tell we will need to either:
1. Remove those additional exports and the associated tests, relying instead on the `process.env.NETWORK` variable (accessed through `config.get()`) to use the library in testnet mode, or
2. keep the `Mainnet` and `Testnet` variants while implementing a new way to create them without the use of the injector

Let me know what you think the best approach is and I can update the branch.

### CODE REVIEW

**Comment by ryanxcharles** [comment]
Hey @ty-everett,

Amazing work. Thanks for taking this on.

First, brief background on the injecter: This tool makes sense if we constantly need to inject dependencies. Originally, this code was going to a be a full node whereby we would need to inject dependencies for the database for testing and other reasons. That is no longer the case, and the only thing the injecter is used for is testnet. That is a really minor issue for the entire codebase to be wrapped with the injecter.

So we should get rid of it.

The only issue is that now we need a new way to distinguish between mainnet and testnet (and regtest and STN) as you state.

My opinion is this:

Classes like Address should have a property that distinguishes the network (Address already has the version byte - other classes have something else). We *should* use an environment variable to automatically set the default. The reason for this is that it is kind of irritating to have to manually set the network everywhere, especially if you already have an environment variable like we do at MB.

However, the network should have the ability to be overridden, and should support both networks simultaneously.

To make this explicit, Address should look something like this:

```
constructor (versionByteNum = defaultVersionByteFor(process.NETWORK))
```

Then there should be a function that gets the network:

```
getNetwork ()
```

^ returns either 'mainnet' or 'testnet' or 'STN' or 'regtest' or whatever.

Then there should be a function that sets the network:

```
setNetwork(networkStr)
```

If `networkStr` is mainnet then it sets the version byte for mainnet, or if it is testnet then it sets the version byte for testnet.

Please let me know your thoughts. I am flexible if you have a better way.

The idea of keeping Testnet classes around is another approach, but sounds a bit more cumbersome to deal with to me. It sounds simpler to just have one class for Address and Bip32. It is easier to understand. Then again, having separate classes would indeed let someone distinguish the type using typeof, which is an advantage.

**Comment by ty-everett** [comment]
Thanks for the feedback, I'll have some more time to put into this hopefully later today.

I like the approach you described in your last paragraph the most because of its simplicity. Right now, users can access a version of the `Address` class that is specific to `Testnet` or `Mainnet`.

My idea is to keep the base `Address` class, which will use the default as it does now. Then to create a few classes that extend address:

```js
const Address = require('./Address')

class TestnetAddress extends Address {
  constructor() {
    super(TESTNET)
  }
}

module.exports = TestnetAddress
```

These extension classes would only be around 10 lines each and would keep the base classes simple. Developers could then just use `TestnetAddress` and `MainnetAddress` in their code if they want to have both at once.

I'll work on an implementation in this branch tonight and get the tests working again.

**Comment by ryanxcharles** [comment]
That makes sense. "Address" can be whatever is the default, and MainnetAddress and TestnetAddress can be extensions that derive from and override the default.

**Comment by ty-everett** [comment]
OK—I've pushed the changes and almost everything should now be working!

Since we've got those new files, I was able to do a cool trick and keep the exact same interface as before, which means none of the test cases had tochange. The `Addres.Mainnet` and `Address.Testnet` variables are exposed as before and they just require the appropriate `MainnetAddress` and `TestnetAddress` classes, respectively.

Since this pattern doesn't _need_ to incorporate non-default dependencies into the classes, this also solves the bundle size problems with adding new `Bip39` language wordlists—I kept the existing exports at the bottom of `bip-39.js` but I left a comment that they should probably be removed (it would just require tests and implementations to change).

The only corner case I haven't managed to solve is the `Kdf` class, as it has no constructor and is really acting as an object with functions as properties that's just kind of _pretending_ to be a class. I'm not sure how important that is or if it's used very often, but I couldn't think of a good way to solve it without changing the interface.

In my view, bundling AES, PBKDF2 and the like in a BitCoin library might not be necessary, because that specific cryptography isn't used or required by the protocol. If people need that functionality, there are many other libraries that can support it. I think that slimming this down and keeping it as lean as possible could be a good design and implementational goal.

So—8854 tests pass, 1 `Kdf` test fails. Let me know what you think of the implementation and what I should do with `Kdf`. I've messed with it for a bit but can't seem to find a good enough solution.

**Comment by ty-everett** [comment]
I went ahead and pushed another commit removing the `kdf` class which got rid of the issues we were having. Since BIP-39 specification says that PBKDF2 is to be used, I think it's safe to rely on the `pbkdf2-compat` dependency directly rather than to use an abstraction.

This makes all tests pass, so I think this is now ready to be merged. Let me know if you think there are other changes that should be made.

**Comment by ryanxcharles** [comment]
Wow, amazing! Great job! I will take a look today if possible.

File: lib/bip-39.js [hunk 1]
```diff
@@ -15,281 +15,278 @@
...
+  This will ensure that whenever new Bip39 languages are added in the future,
+  all of them are always included in bundles that import Bip39. It may be
+  better to simply require users to import lib/bip-39-xx when they wish to use
+  a non-default language.
+
+  it would also be nice to be able to inject a default Bip39 language, similar
+  to how we inject the default NETWORK variable.
+
+  Additionally, Removal of these exports would allow as many new languages as
+  desired to be added without unduly impacting final build output size.
```
Comment (ryanxcharles) [inline lib/bip-39.js:289]: I agree with your opinion. We should not export all languages here. It does seem like simply having a separate class for each language may be useful. That way people can use which ever one they want for which language. That would allow us to add more in a straightforward way without requiring that all users of this class get all of them.

Comment (ryanxcharles) [reply lib/bip-39.js:289]: ...however, that seems like that belongs in a separate PR.

File: lib/bsm.js [hunk 2]
```diff
@@ -2,149 +2,134 @@
...
-let Bsm = inject()
-Bsm.Mainnet = inject({
-  KeyPair: require('./key-pair').Mainnet
-})
-Bsm.Testnet = inject({
-  KeyPair: require('./key-pair').Testnet
-})
+Bsm.magicBytes = Buffer.from('Bitcoin Signed Message:\n')
+
+// TODO: Discuss whether this needs to be supported, it is not used anywhere
```
Comment (ryanxcharles) [inline lib/bsm.js:127]: It seems fine to me to get rid of this. We don't need to propagate complexity around testnet everywhere. If someone actually needs testnet BSM, we can add it back in.

File: lib/ecdsa.js [hunk 3]
```diff
@@ -180,379 +165,376 @@ let inject = function (deps) {
...
 
-inject = require('./injecter')(inject, dependencies)
-let Ecdsa = inject()
-Ecdsa.Mainnet = inject({
-  KeyPair: require('./key-pair').Mainnet
-})
-Ecdsa.Testnet = inject({
-  KeyPair: require('./key-pair').Testnet
-})
+// TODO: Discuss whether this needs to be supported, it is not used anywhere
```
Comment (ryanxcharles) [inline lib/ecdsa.js:532]: LGTM

File: lib/bip-39.js (hunk 1 above)
Comment (ryanxcharles) [reply lib/bip-39.js:289]: Actually, I didn't realize that's how it already works. There already are separate classes. Yeah, let's remove this.
